cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/WorkflowStatsSlaHttpHandler.java

```diff
@@ -0,0 +1,248 @@
+  public void workflowStats(HttpRequest request, HttpResponder responder,
+                            @PathParam("namespace-id") String namespaceId,
+                            @PathParam("app-id") String appId,
+                            @PathParam("workflow-id") String workflowId,
+                            @QueryParam("start") String start,
+                            @QueryParam("end") String end,
+                            @QueryParam("percentile") List<Double> percentiles) throws Exception {
+
+    long startTime = TimeMathParser.parseTimeInSeconds(start);
+    long endTime = TimeMathParser.parseTimeInSeconds(end);
+
+    if (endTime < startTime || startTime < 0 || endTime < 0) {
+      throw new BadRequestException("Wrong start or end time.");
+    }
+
+    for (double i : percentiles) {
+      if (i < 0 || i > 100) {
+        throw new BadRequestException("Percentile values have to be greater than 0 and less than 100");
+      }
+    }
+
+    List<WorkflowDataset.WorkflowRunRecord> workflowRunRecords;
+
+    Id.Program programId = Id.Program.from(namespaceId, appId, ProgramType.WORKFLOW, workflowId);
+    try {
+      try {
+        workflowRunRecords = store.getWorkflowRuns(programId, startTime, endTime);
+
+      } catch (IllegalArgumentException exception) {
+        responder.sendJson(HttpResponseStatus.BAD_REQUEST, "Bad request");
+        return;
+      }
+    } catch (SecurityException exception) {
+      responder.sendJson(HttpResponseStatus.BAD_REQUEST, "You're not allowed to view this");
+      return;
+    }
+
+    int count = workflowRunRecords.size();
+
+    if (count == 0) {
+      responder.sendJson(HttpResponseStatus.OK, "There were no completed runs for this workflow.");
+      return;
+    }
+
+    long average = 0;
+    for (WorkflowDataset.WorkflowRunRecord workflowRunRecord: workflowRunRecords) {
+      average += workflowRunRecord.getTimeTaken();
+    }
+    average /= count;
+
+    workflowRunRecords = sort(workflowRunRecords);
+
+    Map<String, Long> percentileToTime = Maps.newHashMap();
+    Map<String, List<String>> percentileToRunids = Maps.newHashMap();
+    for (double i : percentiles) {
+      List<String> percentileRun = new ArrayList();
+      int percentileStart = (int) ((i * count) / 100);
+      for (int j = percentileStart; j < count; j++) {
+        percentileRun.add(workflowRunRecords.get(j).getWorkflowRunId());
+      }
+      percentileToRunids.put(Double.toString(i), percentileRun);
+      percentileToTime.put(Double.toString(i), workflowRunRecords.get(percentileStart).getTimeTaken());
+    }
+
+    Map<String, List<Long>> actionToRunRecord = getActionRuns(workflowRunRecords);
+
+    Map<String, Map<String, Long>> actionToPercentile = Maps.newHashMap();
+    for (Map.Entry<String, List<Long>> entry : actionToRunRecord.entrySet()) {
+      long mean = 0;
+      for (long iterator: entry.getValue()) {
+        mean += iterator;
+      }
+      mean /= entry.getValue().size();
```

Comment:
again, why the mean is a long, not double? Also, be consistent with variable naming (you call it average above, but mean here).